src/lepy/lego.go

```diff
@@ -0,0 +1,178 @@
+		os.Setenv(k, v)
+	}
+	certificate, err := requestCertificate(CLIArgs.Email, CLIArgs.Server, CLIArgs.CSR, CLIArgs.Plugin)
+	if err != nil {
+		return C.CString(fmt.Sprint("error: couldn't request certificate: ", err))
+	}
+	response_json, err := json.Marshal(certificate)
+	if err != nil {
+		return C.CString(fmt.Sprint("error: coudn't build response message: ", err))
+	}
+	return_message_ptr := C.CString(string(response_json))
+	return return_message_ptr
+}
+
+func requestCertificate(email, server, csr, plugin string) (*LegoOutputResponse, error) {
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't generate priv key: %s", err)
+	}
+	user := LetsEncryptUser{
+		Email: email,
+		key:   privateKey,
+	}
+	config := lego.NewConfig(&user)
+
+	// This CA URL is configured for a local dev instance of Boulder running in Docker in a VM.
+	config.CADirURL = server
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	// A client facilitates communication with the CA server.
+	client, err := lego.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create lego client: %s", err)
+	}
+
+	err = configureClientChallenges(client, plugin)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't configure client challenges: %s", err)
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't register user: %s", err)
+	}
+	user.Registration = reg
+
+	block, _ := pem.Decode([]byte(csr))
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		return nil, errors.New("failed to decode PEM block containing certificate request")
+	}
+	csrObject, err := x509.ParseCertificateRequest(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate request: %s", err)
+	}
+	request := certificate.ObtainForCSRRequest{
+		CSR:    csrObject,
+		Bundle: true,
+	}
+	certificates, err := client.Certificate.ObtainForCSR(request)
+	if err != nil {
+		return nil, fmt.Errorf("coudn't obtain cert: %s", err)
+	}
+
+	return &LegoOutputResponse{
+		CSR:               string(certificates.CSR),
+		PrivateKey:        string(certificates.PrivateKey),
+		Certificate:       string(certificates.Certificate),
+		IssuerCertificate: string(certificates.IssuerCertificate),
+		Metadata: Metadata{
+			StableURL: certificates.CertStableURL,
+			URL:       certificates.CertURL,
+			Domain:    certificates.Domain,
+		},
+	}, nil
+}
+
+func configureClientChallenges(client *lego.Client, plugin string) error {
+	switch plugin {
+	case "http":
+		err := client.Challenge.SetHTTP01Provider(http01.NewProviderServer(os.Getenv("HTTP01_URL"), os.Getenv("HTTP01_PORT")))
+		if err != nil {
+			return errors.Join(errors.New("couldn't set http01 provider server"), err)
+		}
+		err = client.Challenge.SetTLSALPN01Provider(tlsalpn01.NewProviderServer(os.Getenv("TLSALPN01_URL"), os.Getenv("TLSALPN01_PORT")))
```

Comment:
Also not sure if we should keep the ports at 5002 and 5001 respectively. These are the default ports of letsencrypt/pebble, so its convenient for the integration test but if someone actually wants to use it they will have to change their webservers from port 80 and 443.

I tried loading it through the environment variable. Please check out the new readme.